vllm/worker/model_runner_base.py

```diff
@@ -0,0 +1,185 @@
+                                    **kwargs) -> Dict[str, Any]:
+    """
+    Helper method to initialize AttentionMetadata based on an
+    AttentionBackend and broadcastable AttentionMetadata fields.
+    """
+    # Extract the fields used to create AttentionMetadata.
+    valid_attn_kwargs = {}
+    for field in dataclasses.fields(attn_backend.get_metadata_cls()):
+        val = kwargs.pop(field.name, None)
+        if val is not None:
+            valid_attn_kwargs[field.name] = val
+
+    attn_metadata = attn_backend.make_metadata(**valid_attn_kwargs)
+    kwargs["attn_metadata"] = attn_metadata
+    return kwargs
+
+
+def _init_sampling_metadata_from_kwargs(  # type: ignore
+        selected_token_indices: torch.Tensor = None,
+        **kwargs) -> Dict[str, Any]:
+    """
+    Helper method to initialize SamplingMetadata based on broadcastable
+    SamplingMetadata fields.
+    """
+    from vllm.model_executor import SamplingMetadata
+
+    # An empty SamplingMetadata to signal that the worker should skip
+    # sampling.
+    sampling_metadata = SamplingMetadata(
+        seq_groups=None,
+        selected_token_indices=selected_token_indices,
+        categorized_sample_indices=None,
+        num_prompts=0,
+    )
+    kwargs["sampling_metadata"] = sampling_metadata
+    return kwargs
+
+
+def _add_sampling_metadata_broadcastable_dict(
+        tensor_dict: Dict[str, Union[int, torch.Tensor]],
+        sampling_metadata: Optional["SamplingMetadata"]) -> None:
+    """
+    Helper method to update tensor_dict with broadcastable
+    SamplingMetadata fields.
+    """
+    if sampling_metadata is not None:
+        tensor_dict["selected_token_indices"] = (
+            sampling_metadata.selected_token_indices)
+
+
+def _filter_valid_kwargs(cls, kwargs: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Helper method to filter the given kwargs to kwargs that
+    are valid for the given dataclass `cls`.
+    """
+    init_kwargs = {}
+    for field in dataclasses.fields(cls):
+        val = kwargs.get(field.name, None)
+        if val is not None:
+            init_kwargs[field.name] = val
+    return init_kwargs
+
+
+@dataclasses.dataclass(frozen=True)
+class ModelRunnerInputBase(ABC):
+    """Local inputs to each worker's model runner. May contain
+    device-specific data. Different worker backends may have different methods
+    of converting from the global ExecuteModelRequest produced by the LLM
+    engine to the worker-local ModelRunnerInputBase objects.
+
+    Model runners that support multi-GPU execution should define a
+    ModelRunnerInputBase subclass, add their required fields, and specify how to
+    serialize/deserialize a ModelInput for broadcast between workers.
+    """
+
+    @classmethod
+    @abstractmethod
+    def new(cls: Type[T], **kwargs) -> T:
+        """
+        Create a new instance of this class. Populate the new instance with
+        the given kwargs.
+        """
+        raise NotImplementedError
```

Comment:
why do we create a `new` method, rather than directly using `__init__`?

Reply:
The same broadcasted kwargs are deserialized to both WorkerInput and ModelInput dataclasses, so we need to customize the `__init__` to only take the kwargs that are valid for the current dataclass.